### DISCUSSION TITLE
ENDUPRIGHT and ENDDOWNRIGHT recognition

### DISCUSSION THREAD

**Discussion by kystyn** [rdkit/rdkit#5176] [Development]
Hello! I am doing task on recognition of molecula structure by image. And I found that some bonds have bond dir ENDUPRIGHT or ENDDOWNRIGHT. I analyzed my dataset and realized that these bonds are always connected to double bonds. But not all bonds which connected to double bonds are ENDUPRIGHT or ENDDOWNRIGHT. Also, I cannot find law which describes whether to assign ENDUPRIGHT or ENDDOWNRIGHT.
I found this source file (https://github.com/connorcoley/rdchiral/blob/master/rdchiral/bonds.py) and understood that END...-bonds are connected with chirality in some way.
Could you explain me this, please?
Thank you!

**Comment by wopozka** [comment]
ENDUPRIGHT, ENDDOWNRIGHT has nothing to do with chirality, but as you observed with double bond geometry (E/Z or cis/trans). It tells you in which direction the end of a bond points. Imagine Z butene, and lets draw it from left to right. If you analyze the structure, you firts see C-C bond then C=C bond and then another C-C bond. So the end of the first single bond is ENDDOWNRIGHT, as it starts above double bond and ends at down right direction at double bond. Then another single bond C-C is ENDUPRIGHT, as it starts from double bond and points to up right direction. For fully definition of double bond geometry you need to define only two single bonds, therefore it is not necesary to define them for all.

**Reply by kystyn** [reply]
@wopozka Thanks! And how can I understand if single bond starts or ends at double bond by viewing at molecula image?

**Reply by kystyn** [reply]
In the picture below all END...-bonds have ENDUPRIGHT dir
![image](https://user-images.githubusercontent.com/48070175/162277850-22231c4a-3512-42c5-8008-8d502a726d26.png)
And in the following picture bond between double bond and oxygen atom is ENDUPRIGHT and the bond between two double bonds is ENDDOWNRIGHT.
In the same time, all bonds, connected to the leftist double bond are NONE and I cannot catch it
![image](https://user-images.githubusercontent.com/48070175/162278106-f46813e8-bdc7-4037-ba55-e8713fc439bf.png)

**Reply by wopozka** [reply]
@kystyn 
How did you get this molecules, I mean how ENDUPRIGHT, ENDDOWNRIGHT was applied? Was it done by external software and you are trying to analyse it now? It might not be right. 
Anyway, ENDUPRIGHT, ENDDOWNRIGHT depends also on bond direction - whether bond starts at double bond or end at double bond (in rdkit bond have direction from atom to atom). Lets consider bond CH3-C. If bond direction id from CH3 to double bond, then ENDDOWRIGHT should be set, if bond direction is opposite then ENDUPRIGT would be set. The same situation would be for pyrazole-C bond. If bond direction is pyrazole-C then ENDUPRIGHT would be set, and if direction is oposite ENDDOWNRIGHT will be set. The same applies for NN bond. 
For the second molecule the left double bond does have NONE because ZE izomerism is not possible for this bond, the RHS bond would follow the same rules presented above. 
I know it is quite tricky to get, but the best way is first set analyse molecule in one direction, assign bond direction and proper ENDDOWNRIGHT/ENDUPRIGHT parameters, and then check real bond direction. In case it is opposite than you assigned at the begining invert ENDDOWNRIGHT/ENDUPRIGHT.
AFAIR the easiet way get useful information is to use function Chem.Assi nStereoChemistry, that sets proper STEREOZ STEREOE for given double bond. Then you do not need to bother on ENDDOWNRIGHT/ENDUPRIGHT anymore.

**Reply by kystyn** [reply]
@wopozka
How can I understand if ZE isomerism is possible by picture analyzing?

**Reply by wopozka** [reply]
@kystyn Sorry, but it is basic organic chemistry, please read about geometry of double bond and Z/E or cis/trans isomerism.

**Reply by kystyn** [reply]
@wopozka 
Sorry, I forgot to answer your first question.
I build molecules from InChI using rdkit.
I am not chemist, just a software engineer, so I apologize for stupid questions.
Thank you for detailed answer! I hope it will help me :)

**Reply by wopozka** [reply]
@kystyn 
If you are not chemist, or at least does not have at least basic chemical knowledge it might be a challenge. Anyway, if you have double bond you need information about relative position of two substituents on both side of a bond (to other are automatically defined then). If both substituents are pointing in the same direction then you say they are in cis (or Z) position, and if they point in opposite direction then you define them as trans (E) position. This only applies when substituents localized on one side of a bond are different if there are the same cis/trans isomerism does not exist and you do not need to define them. In your case there are to methyl groups on one side of double bond, then it does not make any sense to define anything.

**Reply by greglandrum** [reply]
@kystyn 
> I build molecules from InChI using rdkit. I am not chemist, just a software engineer, so I apologize for stupid questions. 

Maybe we can start with you explaining exactly what you are trying to do.

If the goal is to recognize a molecule image and properly set the stereochemistry of the double bonds, then you just need to pick on substituent on each side of the bond, decide whether or not they are the same side of the double bond, and then set those as the stereo atoms on the bond and set the bond stereo to STEREO_CIS or STEREO_TRANS.

There's a intro to CIS and TRANS on the wikipedia: https://en.wikipedia.org/wiki/Cis%E2%80%93trans_isomerism (that discusses multiple topics, but double bond stereo is the only thing which is relevant here).

**Reply by kystyn** [reply]
@greglandrum 
Hi! Yes, the goal is to properly set stereochemistry of double bonds.
So if I set STEREO_CIS/STEREO_TRANS or STEREOE/STEREOZ  and then call AssignStereo -- will the InChI of recognized molecule be equal to original InChI?
